cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/OptimizationOptions.java

```diff
@@ -20,20 +20,7 @@
   private final boolean _isTriggeredByGoalViolation;
   private final Set<Integer> _requestedDestinationBrokerIds;
   private final boolean _onlyMoveImmigrantReplicas;
-
-  /**
-   * Default value for {@link #_excludedBrokersForLeadership} is an empty set.
-   */
-  public OptimizationOptions(Set<String> excludedTopics) {
-    this(excludedTopics, Collections.emptySet());
-  }
-
-  /**
-   * Default value for {@link #_excludedBrokersForReplicaMove} is an empty set.
-   */
-  public OptimizationOptions(Set<String> excludedTopics, Set<Integer> excludedBrokersForLeadership) {
-    this(excludedTopics, excludedBrokersForLeadership, Collections.emptySet());
-  }
```

Comment:
Just curious, is the reason to remove these two method because both the 2nd and 3rd parameters are Set<int>, which causes confusion?

Reply:
This class uses _telescoping constructor pattern_.  It does not scale well with the growing number of parameters, because it gets harder to write client code and read it. As you correctly identified 2nd and 3rd parameters are providing a similar functionality (i.e. both `Set<Integer> excludedBrokersForX`). Hence, having a constructor with the first one, but not the other indeed causes confusion. That is why these two constructors were dropped.

That being said, there are alternatives to this design pattern. In particular _Builder pattern_ would be a good alternative (see `ProvisionRecommendation` for an example use of it). The challenge with builder pattern is that the inheritance becomes more messy -- i.e. requires all the non-leaf classes to be abstract -- and `OptimizationOptions` is a class that we would like to extend from (internally).